Source/JavaScriptCore/interpreter/Interpreter.cpp

```diff
@@ -138,7 +147,7 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
     auto programSource = programString->value(globalObject);
     RETURN_IF_EXCEPTION(scope, JSValue());
 
-    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::DirectEval, programSource, program)) {
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !isTrusted && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::DirectEval, programSource, *vm.emptyList)) {
```

Comment:
This `canCompileStrings` does not have exception check.

Reply:
If i put a RETURN_IF_EXCEPTION isnide of the if block it fails the tests because it starts returning a typeerror rather than the required eval.

I have put a `scope.clearException();` inside instead based on similar code in the unsafe-eval check in FunctionConstructor.cpp

Reply:
Hmm, this sounds strange. Swallowing the already thrown error is not a good thing.
Why is TypeError thrown from the first place? Can you replace that TypeError with EvalError when it is necessary?

Reply:
The default policy can also end up throwing arbitrary exception types. Normally these TypeErrors and arbitrary userland exceptions would be propogated up but for eval and Function they all need to be swallowed and replaced with an EvalError.

I'm happy to move the scope.clearExceptions down into canCompileStrings and return a false for when an exception exists rather than propagating up to here to be cleared?

Reply:
Can you specify the part of the spec corresponding to this part?

> The default policy can also end up throwing arbitrary exception types. Normally these TypeErrors and arbitrary userland exceptions would be propogated up but for eval and Function they all need to be swallowed and replaced with an EvalError.

Reply:
I've rejigged the code a bit and added the RETURN_IF_EXCEPTION checks inside of the if statement and afterwards to catch both code branches

Reply:
Ditto.